lib/nixos-system.nix

```diff
@@ -1,43 +1,93 @@
-lib.nixosSystem (lib.recursiveUpdate (removeAttrs args [ "globalSpecialArgs" "nixosSpecialArgs" "hmSpecialArgs" ]) {
+with lib;
+
+{
+  # The global modules are included in both NixOS and home-manager.
+  globalModules ? [ ]
+
+  # Home-manager specific modules.
+, hmModules ? [ ]
+
+  # NixOS specific modules.
+, nixosModules ? [ ]
+
+  # The global extra arguments are included in both NixOS and home-manager.
+, globalSpecialArgs ? { }
+
+  # Home-manager specific extra arguments.
+, hmSpecialArgs ? { }
+
+  # NixOS specific extra arguments.
+, nixosSpecialArgs ? { }
+
+, ...
+} @ args:
+let
+  args' = removeAttrs args [
+    "globalModules"
+    "nixosModules"
+    "hmModules"
+
+    "globalSpecialArgs"
+    "nixosSpecialArgs"
+    "hmSpecialArgs"
+  ];
+in
+nixosSystem (recursiveUpdate args' {
   specialArgs = {
-    mode = "NixOS";
-    soxin = self;
-  } // globalSpecialArgs // nixosSpecialArgs;
+    # send home-manager down to the NixOS modules
+    inherit home-manager;
 
-  modules = modules ++ [
-    {
-      _module.args = { inherit home-manager; };
-    }
+    # the mode allows us to tell at what level we are within the modules.
+    mode = "NixOS";
 
-    self.nixosModules.soxin
+    # send soxin to all NixOS modules
+    soxin = self;
+  }
+  # include the global special arguments.
+  // globalSpecialArgs
+  # include the NixOS special arguments.
+  // nixosSpecialArgs;
 
-    home-manager.nixosModules.home-manager
-    # Required when using flakes.
-    {
+  modules =
+    # include the global modules
+    globalModules
+    # include the NixOS modules
+    ++ nixosModules
+    # include all Soxin modules
+    ++ (builtins.attrValues self.nixosModules)
+    # include all home-manager modules
+    ++ (builtins.attrValues home-manager.nixosModules)
+    # configure Nix registry so users can find soxin
+    ++ singleton { nix.registry.soxin.flake = self; }
+    # configure home-manager
+    ++ (singleton {
+      # tell home-manager to use the global (as in NixOS system-level) pkgs and
+      # install all  user packages through the users.users.<name>.packages.
       home-manager.useGlobalPkgs = true;
       home-manager.useUserPackages = true;
-    }
-
-    # Override home-manager per-user submodule to add our own modules to it,
-    # and to pass the argument `mode`.
-    ({ config, ... }: {
-      options.home-manager.users = lib.mkOption {
-        type = lib.types.attrsOf (lib.types.submoduleWith {
-          modules = [
-            {
-              _module.args = { inherit home-manager; };
-            }
-            self.nixosModules.soxin
-          ];
-          specialArgs = {
-            mode = "home-manager";
-            soxin = self;
-          } // globalSpecialArgs // hmSpecialArgs;
-        });
-      };
-    })
-  ];
+
+      home-manager.extraSpecialArgs = {
+        # send home-manager down to the home-manager modules
+        inherit home-manager;
```

Comment:
Same as above.

Reply:
Resolving this to the discussion centralized.